notebooks/rocoto.ipynb

```diff
@@ -0,0 +1,1013 @@
+      "[2024-10-28T12:34:55]    ERROR Error at workflow -> tasks -> task_greet:\n",
+      "[2024-10-28T12:34:55]    ERROR   'command' is a required property\n",
+      "[2024-10-28T12:34:55]    ERROR Error at workflow:\n",
+      "[2024-10-28T12:34:55]    ERROR   'log' is a required property\n"
+     ]
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "YAML validation errors\n"
+     ]
+    }
+   ],
+   "source": [
+    "try:\n",
+    "    rocoto.realize(\n",
+    "        config=Path('fixtures/rocoto/err-workflow.yaml'),\n",
+    "        output_file='tmp/err-workflow.xml'\n",
+    "    )\n",
+    "except Exception as e:\n",
+    "    print(e)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "cbad7aef-02b6-43a3-9241-349affa2f71c",
+   "metadata": {},
+   "source": [
+    "### Entities and Cyclestrings\n",
+    "\n",
+    "Constants called entities may be defined so that their values can be referenced throughout the rest of the workflow. These are defined in an `entities:` block, with their names and values given as keys and values in the YAML. Below an entity named `LOG` is defined with a string value. This value is referred to in the other parts of the workflow with the syntax `&ENTITY_NAME;`. In this case, note the `&LOG;` entity within the `log:` block.\n",
+    "<!--cell 14-->"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "id": "074a6e7e-7c05-4037-b954-06eba8ae2241",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "workflow:\n",
+      "  attrs:\n",
+      "    realtime: false\n",
+      "    scheduler: slurm\n",
+      "  cycledef:\n",
+      "    - spec: 202410290000 202410300000 06:00:00\n",
+      "  entities:\n",
+      "    LOG: \"2024-10-29/test06:00:00.log\"\n",
+      "  log: logs/&LOG;\n",
+      "  tasks:\n",
+      "    task_greet:\n",
+      "      command: echo Hello, World!\n",
+      "      cores: 1\n",
+      "      walltime: 00:00:10\n"
+     ]
+    }
+   ],
+   "source": [
+    "%%bash\n",
+    "cat fixtures/rocoto/ent-workflow.yaml"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "34cbb0e1-ac5f-4764-afbb-9574966542c6",
+   "metadata": {},
+   "source": [
+    "Cycle strings represent dynamic cycle time components that are represented by specific flags and are rendered at run time. Here, the `LOG` entity contains `@Y`, `@m`, `@d` and `@X` flags that represent the current year, month, day, and time respectively. For more information on these flags, see <a href=\"https://christopherwharrop.github.io/rocoto/\">Rocoto Documentation</a>. A `cyclestr:` block is used to mark a string containing cycle string flags for rendering at runtime. Here, since the `LOG` entity contains these flags, a `cyclestr:` block within the `log:` block marks the string for rendering at runtime. This string itself is contained in a `value` key.\n",
```

Comment:
```suggestion
    "Cycle strings represent dynamic cycle time components that are represented by specific flags and are rendered when Rocoto runs the XML. Here, the `LOG` entity contains `@Y`, `@m`, `@d` and `@X` flags that represent the year, month, day, and time relative to a cycle defined by the `cycledefs:` entry. For more information on these flags, see <a href=\"https://christopherwharrop.github.io/rocoto/\">Rocoto Documentation</a>. A `cyclestr:` block is used to mark a string containing cycle string flags for rendering when Rocoto runs. Here, since the `LOG` entity contains these flags, a `cyclestr:` block within the `log:` block indicates that the flags should be rendered when Rocoto runs. This string itself is contained in a `value` key.\n",
```